### ISSUE TITLE
Real time face recognition with keras

### ISSUE THREAD

**Issue by aquibjaved** [fchollet/keras#4091]
I'am doing project on real time face recognition, I found a code on github which uses Yale dataset and I got accuracy of 89%, I've changed the dataset and save the weight. Now I want to make it real time, So now I load an Image and convert it to numpy array then reshape it and pass it though **classes = loaded_model.predict_classes(img)**  
Now here I am getting an error which says : **Exception: Error when checking : expected dense_input_1 to have shape (None, 22500) but got array with shape (22500, 1)**

I just want to know that in **predict_classes** function what type of data feed required a numpy array 1d or 2d ?

**Comment by nixingyang** [comment]
Hi,
I think the shape of the variable *img* is incorrect. You may check the [expand_dims](https://docs.scipy.org/doc/numpy/reference/generated/numpy.expand_dims.html) method and add an additional axis at the beginning.
By the way, do you plan to pass single image to the model at each time? This might be inefficient. You may try different batch size and verify the relationship between computational speed and batch size.
BR.

**Comment by aquibjaved** [comment]
Hi, nixingyang,
Thank for you reply. I've tried to do, but facing the same problem...below I've written my last lines of codes. please help me out ... 

> img = cv2.imread('image.jpg',0)
  img = cv2.resize(img,(150,150))
  img = np.expand_dims(img, axis=0)
  img = np.array(img).flatten()
 
  classes = loaded_model.predict_classes(img)

**Comment by nixingyang** [comment]
@aquibjaved Could you please post the code which defines the structure of your model? If possible, the output of **model.summary()** would also be helpful.

**Comment by aquibjaved** [comment]
@nixingyang  below is the code to load the model and an Image for test.
`import cv2
import numpy as np
import os
from skimage import io
from sklearn.cross_validation import train_test_split


#importing kers
from keras.utils import np_utils
from keras.models import *
from keras.layers import Dense, Dropout, Activation

# load json and create model
json_file = open('model.json', 'r')
loaded_model_json = json_file.read()
json_file.close()
loaded_model = model_from_json(loaded_model_json)

# load weights into new model
loaded_model.load_weights("yaleface.h5")
print("Loaded model from disk")


# evaluate loaded model on test data
loaded_model.compile(loss='binary_crossentropy', optimizer='rmsprop', metrics=['accuracy'])
#score = loaded_model.evaluate(X_test,Y_test, verbose=0)
print loaded_model.summary()


img = cv2.imread('image.jpg',0)

img = cv2.resize(img,(150,150))

img = np.expand_dims(img, axis=0)
img = np.array(img).flatten()

#img = np.expand_dims(np.expand_dims(img, axis=0), axis=0)/255
print img
classes = loaded_model.predict_classes(img)
`

**Comment by nixingyang** [comment]
@aquibjaved The structure of your model was not revealed in your code as you initiated the model from a json file. Could you please try the code shown below?

```python
img = cv2.imread('image.jpg', 0).astype(np.float32) / 255
img = cv2.resize(img, (150, 150))
img = img.flatten()
img = np.expand_dims(img, axis=0)
```

The shape of the input would be (1, 22500) which should work in your case. The weird thing is that, you intended to apply **flatten** operation on your input. It smashed the spatial dimensions of the image.

**Comment by aquibjaved** [comment]
@nixingyang 

Thank you so much, It works but for every different image it says 
classes predict is [4].
it's weird, How can I overcome this problem

**Comment by nixingyang** [comment]
@aquibjaved This depends on how good is your model. I noticed a bug in your code. You should call **compile** before **load_weights**. The pre-trained weights were lost when you called **compile**.

**Comment by aquibjaved** [comment]
@nixingyang I did, I keep the **compile**  before ** load_weights**, same issue


`import cv2
import numpy as np
import os
from skimage import io
from sklearn.cross_validation import train_test_split

#importing kers
from keras.utils import np_utils
from keras.models import *
from keras.layers import Dense, Dropout, Activation

# load json and create model
json_file = open('model.json', 'r')
loaded_model_json = json_file.read()
json_file.close()


loaded_model = model_from_json(loaded_model_json)
loaded_model.compile(loss='binary_crossentropy', optimizer='rmsprop', metrics=['accuracy'])

# load weights into new model
loaded_model.load_weights("yaleface.h5")
print("Loaded model from disk")

#print loaded_model.summary()


img = cv2.imread('test2.jpg', 0).astype(np.float32) / 255
img = cv2.resize(img, (150, 150))
img = img.flatten()
img = np.expand_dims(img, axis=0)

classes = loaded_model.predict_classes(img)
print classes`

**Comment by nixingyang** [comment]
@aquibjaved Comment the **load_weights** line and try again. The predictions should be random for each input image. Then, the issue is related to the pre-trained weights.

**Comment by aquibjaved** [comment]
@nixingyang  I did comment the **load_weight** and the predictions are random, on a same image every time the prediction is different. So now there's an issue related to the pre-trained weights.

**Comment by aquibjaved** [comment]
@nixingyang 
My accuracy on test data is 98.4% then why it is not working ?
![Uploading Screenshot from 2016-10-18 15-24-14.png…]()

**Comment by nixingyang** [comment]
@aquibjaved Make sure that you are using the same version packages as the pre-trained model. Check how the weights are saved/loaded to/from disk. It should be a bug in your code.

**Comment by kretes** [comment]
@aquibjaved have you tried exporting your whole model with weights?

I am using save_model and load_model for me and it is working correctly - i.e. the loaded model has expected metric value.

2nd question: am I right that you are flattening your image before passing it to the net? If so - is it flattening it the same way it was when training. And why not to keep it a 2D network input? Are the values during training somehow preprocessed - do you do the same when calling predict ?

**Comment by aquibjaved** [comment]
@kretes   I'll try this surely... load_model

**Comment by aquibjaved** [comment]
@kretes  can show you'r code how you've loaded the model and made the predictions.
